Added js module: new JavaScript API supporting V8

snej opened the pull request:
Part of https://github.com/couchbase/sync_gateway/pull/5980 -- I just moved the `js` module here, and the `v8go` package dependency.

snej committed:
Added js module: new JavaScript API supporting V8

torcolvin reviewed (commented):
A bunch of pedantic nitpicks, but shouldn't 

js_map_fn.go, js_runner.go, js_server.go

be obsolete with this commit + corresponding SG commit?

torcolvin reviewed (commented):
The second set of changes I'd like to make are:

1.  is to make the code panic-less, even though usually this amounts to programmer error if there's a problem that would case a panic. The reason for this is that we could bring down a SG process with multiple databases, where only would fail in a single database, even if it doesn't start up, or fails to update sync function
2. pass contexts into the functions where possible, since we are using them more and more for logging. Instead of using `context.Background()` replace with `context.TODO()`. Some of these contexts are inside the js code only, but some of these functions I think call back into sync gateway where the context logging could be helpful to us.

snej commented on js/v8_runner.go:
```diff
@@ -0,0 +1,272 @@
+	// Create a V8 Context and run the setup script in it:
+	ctx := v8.NewContext(vm.iso, template.Global())
+	if _, err := vm.setupScript.Run(ctx); err != nil {
+		return nil, errors.Wrap(err, "Unexpected error in JavaScript initialization code")
```
Yup!

snej commented on js/otto_runner.go:
```diff
@@ -0,0 +1,207 @@
+	"fmt"
+	"strconv"
+
+	"github.com/pkg/errors"
```
ok

snej commented on js/otto_vm.go:
```diff
@@ -0,0 +1,118 @@
+		return nil, fmt.Errorf("the js.VM has been closed")
+	}
+	if vm.curRunner != nil {
+		panic("illegal access to v8VM: already has a v8Runner")
```
ok

snej commented on js/otto_vm.go:
```diff
@@ -0,0 +1,118 @@
+		panic("illegal access to v8VM: already has a v8Runner")
+	}
+	if !vm.services.hasService(service) {
+		return nil, fmt.Errorf("unknown js.Service instance passed to VM")
```
ok

snej commented on js/v8_runner.go:
```diff
@@ -0,0 +1,272 @@
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
```
ok

snej commented on js/v8_template.go:
```diff
@@ -0,0 +1,200 @@
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
```
ok

snej commented on js/v8_vm.go:
```diff
@@ -0,0 +1,247 @@
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
```
ok

snej commented on js/v8_utils.go:
```diff
@@ -0,0 +1,134 @@
+// Calls v8.Isolate.ThrowException, with the Go error's string as the message.
+func v8Throw(i *v8.Isolate, err error) *v8.Value {
+	var errStr string
+	// if httpErr, ok := err.(*base.HTTPError); ok {
+	// 	errStr = fmt.Sprintf("[%d] %s", httpErr.Status, httpErr.Message)
+	// } else {
```
deleted.

snej commented on js/v8_vm.go:
ok

snej commented on js/v8_vm.go:
ok

snej commented on js/vmpool_test.go:
```diff
@@ -0,0 +1,219 @@
+
+func TestPoolsConcurrently(t *testing.T) {
+	maxProcs := runtime.GOMAXPROCS(0)
+	log.Printf("FYI, GOMAXPROCS = %d", maxProcs)
```
ok

snej committed:
Addressing review comments

snej committed:
Added license text

snej committed:
js module: Improved use of Contexts

- VM and VMPool now have an associated Context, which must be given when
  creating one. (VMs created by VMPool inherit its Context.)
- VM and VMPool now have a Context() method.
- Runner.Context() now returns the VM's Context by default, but calling
  Runner.SetContext() overrides it until the Runner is returned.
- Removed Runner.ContextOrDefault(); use Context() instead.
- The `ctx` parameter to Service.Run() may be nil, in which case the default
  Context (the VM's) is used.
- Tests use a new testCtx() function that's similar to the one in SG.base.
  It creates a Context derived from context.TODO that cancels when the test
  exits.

Some renaming to avoid confusion:
- Renamed v8Runner.ctx to v8ctx.
- Renamed baseRunner.goContext to ctx.
- Renamed other variables of type *v8.Context to v8ctx.

snej commented:
I've implemented change 2, adding Contexts, in commit c770934. See the commit log for details.

The addlicense check is failing; it needs to be disabled for js/underscore-umd-min.js and js/underscore-umd.js .

snej committed:
Removed panics

- panics in illegal situations replaced by logError() and returning,
  with an error if possible.
- V8Runner.NewInt() and NewString() now return an error because the V8 call
  returns an error, although I believe it can only fail in the unlikely case
  that V8 runs out of memory.

snej commented:
Change 1 (no panics) now implemented in [5ee3866](https://github.com/couchbase/sg-bucket/pull/91/commits/5ee386619ff4b3d56763eade02f97d3506435ac6). This required a few change in sync_gateway, which I've also pushed.

snej committed:
v8go package moved from snej to couchbasedeps

torcolvin committed:
Add v8 CI (#95)

* add v8 tests

* Ignore underscore files

* use quotes
